### ISSUE TITLE
Adapter API - discoverDevices()

### ISSUE THREAD

**Issue by benfrancis** [moziot/gateway#18]
**_As a MozIoT web app developer I would like to be able to request a list of new devices which aren't yet added to the gateway._**

In order for the web app to display a list of devices available on the network which can be added to the gateway, it needs to be able to get a list of devices from the adapter manager.

My suggestion would be:

AdapterManager.discoverDevices()

This method would return a Promise which resolves with a list of devices. E.g.

```
[{
  id: '123e4567-e89b-12d3-a456-426655440000',
  type: 'onOffSwitch',
  name: ' Aeotec Smart Switch 6'
}]
```

- **id** - A UUID generated by the AdapterManager which internally maps this ID onto an (e.g.) Z-Wave or ZigBee identifier.
- **type** - A Web Thing type, as defined in the [Web Thing API](https://moziot.github.io/wot/) specification.
- **name** - A user readable name which may give a hint to the user as to which device it is (e.g. manufacturer name, model number etc.).

My expectation is that behind the scenes this method would iterate through the list of adapters and initiate a protocol specific process for listing available devices (e.g. entering a pairing or commissioning mode). Ideally the actual pairing shouldn't take place with a further user interaction and a separate AdapterManager.pair(id) method.

**Comment by benfrancis** [comment]
@dhylands What do you think?

**Comment by dhylands** [comment]
Sure we can do that. For both ZigBee and ZWave, it will have the entire network cached, so discover will just be returning information from the cache.

How do you want to deal with new devices being added/removed when not initiated by the user? Currently I have the adapter emitting a device-added notification.

Presumably we should be doing something similar for adapters as well? i.e. have an discoverAdapters() which will return a promise which returns a list of adapters? We should probably support adapters being added/removed as well (USB dongles can be plugged in/unplugged). I definitely want to support multiple dongles of each type.

**Comment by dhylands** [comment]
We can have discoverDevices on both the AdapterManager (which will return all devices from all adaptes) and discoverDevices on an Adapter (which will return all devices from a particular adapter).

**Comment by benfrancis** [comment]
@dhylands wrote:
> Sure we can do that. For both ZigBee and ZWave, it will have the entire network cached, so discover will just be returning information from the cache.

For this method I'm more concerned with devices which aren't yet connected to the network. The way that the SmartThings UI works is that you tap "add device" in the app and then press the pairing button on a ZigBee/Z-Wave device and it lists the new devices. I'm looking to do something similar here.

> How do you want to deal with new devices being added/removed when not initiated by the user? Currently I have the adapter emitting a device-added notification.

How can devices be added to the network without being initiated by the user? For the XStick I don't think that's possible because pairing mode has to be activated in software (unless devices are paired when it's plugged into another device before hand and somehow persisted on the I suppose...). For the Z-Stick are you talking about the pairing button on the dongle itself (which requires user interaction), or something else?

Paired devices that have gone out of range of the network or run out of battery should still be listed in the UI, just maybe marked as disconnected at some point. Can devices be un-paired without user interaction..?

Anyway, I suppose discoverDevices() could just return both cached and new device IDs. The web app could compare the list of IDs against the database to figure out what's new. But to be clear the main purpose of this method is to end up with a list of devices which *aren't* yet in the gateway database, so they can be added. To get a list of connected devices the web app will just query the gateway database.

In general for security my expectation is that pairing and unpairing devices should always be initiated by the user in some way.

> Presumably we should be doing something similar for adapters as well? i.e. have an discoverAdapters() which will return a promise which returns a list of adapters? We should probably support adapters being added/removed as well (USB dongles can be plugged in/unplugged). I definitely want to support multiple dongles of each type.

I haven't thought much about this. Presumably a consumer gateway product would have the adapters permanently attached, but the DIY Raspberry Pi case this is certainly relevant. Maybe we should have a list of adapters in Settings.

> We can have discoverDevices on both the AdapterManager (which will return all devices from all adaptes) and discoverDevices on an Adapter (which will return all devices from a particular adapter).

I can't think of a use case in the UI for getting a list of devices connected to just one of the adapters, but sure if you think it's useful.

Sorry if I wasn't clear on some of these points, I've edited the original description a bit to be more clear.

**Comment by dhylands** [comment]
On some dongles, there is a button on the dongle that you can press to initiate pairing mode. I know for sure this works with the dongle I have. So its still user initiated, just not through the software. I know for sure that this works when the dongle is unplugged (the docs say that you have to unplug - I'll do a test later today), I'm not sure if it works when the dongle is plugged in. So devices can definitely be added to the network without using the host software.

> We can have discoverDevices on both the AdapterManager (which will return all devices from all adaptes) and discoverDevices on an Adapter (which will return all devices from a particular adapter).
> 
> I can't think of a use case in the UI for getting a list of devices connected to just one of the adapters, but sure if you think it's useful.

I was thinking more along the lines of the adapter manager's discoverDevices would call each of the Adapter's discoverDevices. This way the adapter manager doesn't also need to keep a cache of devices.

> Can devices be un-paired without user interaction..?

without user interaction in the UI - yes. But the user still needs to initiate by unplugging the dongle and pressing the button (so the security aspect is still maintained).

I think we definitely want to support dynamic adapter removal/addition since the user wants to be able to unplug the dongle, travel to the device and do the pairing. Devices which are temporarily off or asleep still get reported by the dongle since the dongle has a cache of the devices. Some of their details might not be fully available until they wake up. I'm happy to defer dynamic adapter support until later.

The way that zwave works, opening the serial port automatically initiates a scan and all devices are reported to the host at that time. After all of the awake nodes have reported in, then we get a "scan" complete from the zwave dongle. After that, the only way new devices should be added/removed is via pairing (either sw initiated or user initiated by unplugging the dongle and using the button).

So I'm assuming that the app will call discoverDevices and if the scan is in progress then it may get deferred until the scan is complete. Any time discoverDevices is called while the dongle is plugged in and after the scan is complete it will return all of the devices. When you initiate pairing and the user presses the we get a "node-added" notification which we need to pass on to the UI. Any time after the node-added event occurs calling discoverDevices would include this new node.

Is this how you were expecting discoverDevices to work?

**Comment by benfrancis** [comment]
> I was thinking more along the lines of the adapter manager's discoverDevices would call each of the Adapter's discoverDevices. This way the adapter manager doesn't also need to keep a cache of devices.

Oh yeah that's fine.

> The way that zwave works, opening the serial port automatically initiates a scan and all devices are reported to the host at that time. After all of the awake nodes have reported in, then we get a "scan" complete from the zwave dongle.

I assume this is just lists devices which have already been paired? It doesn't just pair with everything around it when you open a serial port?

> After that, the only way new devices should be added/removed is via pairing (either sw initiated or user initiated by unplugging the dongle and using the button).

> Is this how you were expecting discoverDevices to work?

Sort of. I hadn't really anticipated the whole unplugging the dongle and running it off a battery while you wander around the house thing! That seems strange to me, do other Z-Wave dongles provide that feature? I doubt you'd be able to do this with a consumer gateway product.

I think what I imagined was that you could scan for un-paired devices and then choose which one(s) to pair. (Isn't that how Bluetooth works?) It sounds like you can just get a list of already-paired devices, or initiate pairing mode on the controller and pair a device by pushing its button. So the UI I designed where there's Step 1) search for un-paired devices and Step 2) choose which device(s) to pair with probably won't work?

How do you know you're pairing with the device you're intending to pair with, and not a malicious device which happens to be in pairing mode at the same time? That's one of the security risks I'm aware of associated with push-button commissioning.

**Comment by dhylands** [comment]
> Sort of. I hadn't really anticipated the whole unplugging the dongle and running it off a battery while you wander around the house thing! That seems strange to me, do other Z-Wave dongles provide that feature? I doubt you'd be able to do this with a consumer gateway product.
> 
> I think what I imagined was that you could scan for un-paired devices and then choose which one(s) to pair. (Isn't that how Bluetooth works?) It sounds like you can just get a list of already-paired devices, or initiate pairing mode on the controller and pair a device by pushing its button. So the UI I designed where there's Step 1) search for un-paired devices and Step 2) choose which device(s) to pair with probably won't work?

Neither Zigbee nor ZWave really has the notion of scanning for unpaired devices. Bluetooth does.

With ZigBee you tell the master to allow or disallow new nodes to join, then you typically press a button on the device you want to join and it joins the network.

ZWave is similar. You send an addNode command to the dongle, and then you press a button (or some similar procedure) and it automatically joins the network.

With ZWave, the dongle then sends a notification that a new node was added. You get the same information from the node that you get during the initial scan.

With ZWave the scanning portion is all automatic. There is no command to scan the network. With ZigBee you have to scan the network (you ask the master node for its neighbors and its routing table and then recurse). But it still only reports paired nodes.

I expect for ZigBee we'll normally disallow network joins, and then the user would initiate a pair, we enable joins and then after the user presses the button on the device we'll find out about the new node (I'm not sure if there is a specific notification or if we need to rescan)
> 
> How do you know you're pairing with the device you're intending to pair with, and not a malicious device which happens to be in pairing mode at the same time? That's one of the security risks I'm aware of associated with push-button commissioning.

If somebody happened to be nearby and pressed the button on their device before you did, then yeah they could get their device on your network. I'm not aware of anyway of preventing that with ZWave or ZigBee.

**Comment by dhylands** [comment]
Now some devices, like Sonos speakers, Hue Hub, many IP cameras can be scanned and detected before you know usernames/passwords etc.

**Comment by dhylands** [comment]
> So the UI I designed where there's Step 1) search for un-paired devices and Step 2) choose which device(s) to pair with probably won't work?

Actually, now that I'm able to run the app, I think that it's still valid.

But rather than adding "unpaired" devices, perhaps it would addidevices which haven't previously been registered in the DB. This should probably include detetable devices which we don't already know about, but it would also need a way to initiate pairing for each adapter.

So perhaps Clicking on the + will show a list of any devices which exist on the zwave network and which don't exist in the DB, and there would also be a button for each adapter which adds devices to that adapter.

**Comment by benfrancis** [comment]
I'm trying to make the experience of adding new Things as simple as possible so that it's basically 1) tap add in the UI 2) tap a pairing button on the device, done. The end user should not have to even know how many adapters there are in the gateway, or that it supports ZigBee and Z-Wave, as far as they're concerned there are only Web Things.

I think I confused matters by trying to describe a two step process. I tried adding a device with SmartThings again and I've realised that how it actually works is that tapping "Add a Thing" puts the hub in pairing mode so that when you press the pairing button on a device it is immediately paired. All that the second step of tapping "Save" does is to save the custom name you've given to the Thing. If you exit the add device dialog before tapping save the device is still actually paired.

So I suppose at least for ZigBee and Z-Wave the discoverDevices() method I asked for is in reality a turnOnPairingModeOnAllAdapters() method :P

The current implementation of discoverDevices() in https://github.com/moziot/gateway/pull/21 is really for the use case of adding Things to the gateway database which have previously been paired with an adapter using some external mechanism in hardware. In my mind this is actually an edge case which is a quirk of the Z-Wave controller we're using and would probably never happen in a consumer gateway with all the adapters built in internally. It's possible that a consumer gateway might have a physical button for pairing devices (using all radios), but if that was used then I would expect devices to just appear in the UI automatically without having to go in and add them to the UI separately.

So in summary now that I understand it better I think what I actually needed discoverDevices() to do is to put all the radios in pairing mode. It might make sense to rename this to something like pair() instead, except that other adapters like Bluetooth and WiFi are probably going to work differently.

What do you think, @dhylands ?

**Comment by benfrancis** [comment]
discoverDevices() was implemented in https://github.com/moziot/gateway/commit/5c7d041ab771c5c719a256f26b4e86e27fac9278 to get a full list of devices from adapters and addNewDevice() was implemented in https://github.com/moziot/gateway/commit/3b55b2f439e9434dfbe9febdc34fb9bd8c446534 to start pairing on all adapters that support pairing.
